### DISCUSSION TITLE
Import Mount commands from external library.

### DISCUSSION THREAD

**Discussion by Asiratlan** [cypress-io/cypress#27346] [Questions and Help]
Hello there,

I have this custom mount command that wrap a tested component in providers to use Formik, React-Router and React-Redux, this command works properly into my project:

```
Cypress.Commands.add('mountWithFormik',  (component, options = {}) => {
  const {
    routerProps = { initialEntries: ['/'] },
    store = createMockStore(),
    initialValues = {},
    ...mountOptions
  } = options;

  const wrapped = (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <MemoryRouter {...routerProps}>
      <Provider store={store}>
        <Formik initialValues={initialValues}>
          <Form>{component}</Form>
        </Formik>
      </Provider>
    </MemoryRouter>
  ););
```

I have been trying to move this code from the supportFile into a reusable library that the supportFile would then import this way:

```
// eslint-disable-next-line import/no-extraneous-dependencies
import 'my-lib/dist/commands';
```

When I'm trying to execute my test I end up with the following error:

` > Could not find "store" in the context of "Connect(ComponentToTest)". Either wrap the root component in a <Provider>, or pass a custom React context provider to <Provider> and the corresponding React context consumer to Connect(ComponentToTest) in connect options.`.

Is there any reason why the Redux-Provider would not be working the same as if this command was hosted directly into the supportFile? The library should have all the required dependancy installed to work properly.

**Comment by portyu9** [comment]
Hi @Asiratlan 

Moving the command to another package is fine by itself. The likely problem is that the package is bringing/bundling its **own copy of React or `react-redux`**.

`<Provider>` and `connect()` communicate through React Context. If the external library's `<Provider>` comes from one copy of `react-redux`, while `Connect(ComponentToTest)` comes from another copy, they are using different context instances. The component therefore behaves exactly as if there were no Provider:

```text
Could not find "store" in the context of "Connect(ComponentToTest)"
```

React Redux specifically calls out duplicate React and duplicate React Redux installations as causes of context problems.

For a reusable Cypress helper library, packages such as these should normally be peer dependencies rather than bundled private copies:

```json
{
  "peerDependencies": {
    "react": "...",
    "react-dom": "...",
    "react-redux": "..."
  }
}
```

Also configure the library bundler to leave those packages external rather than embedding them in `dist`.

A useful diagnostic in the consuming project is:

```bash
npm ls react react-dom react-redux
```

and inspect your built `my-lib/dist/commands` to make sure React/React Redux have not been bundled into it.

Cypress's current recommendation is still to expose a custom `cy.mount()` from the support setup and wrap it with application providers. That reusable implementation can live in a package; it just needs to share the consumer's React/React Redux instances.

Please mark this answer as accepted if it helped, thank you.
